core/banking/fee_discount.go

```diff
@@ -0,0 +1,162 @@
+// Copyright (C) 2023 Gobalsky Labs Limited
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package banking
+
+import (
+	"context"
+
+	"code.vegaprotocol.io/vega/core/events"
+	"code.vegaprotocol.io/vega/libs/num"
+	"code.vegaprotocol.io/vega/logging"
+)
+
+type partyAssetKey struct {
+	party string
+	asset string
+}
+
+func (e *Engine) feeDiscountKey(asset, party string) partyAssetKey {
+	return partyAssetKey{party: party, asset: asset}
+}
+
+func (e *Engine) RegisterTakerFees(ctx context.Context, assetID string, feesPerParty map[string]*num.Uint) {
+	updateDiscountEvents := make([]events.Event, 0, len(e.feeDiscountPerPartyAndAsset))
+
+	// ensure asset exists
+	asset, err := e.assets.Get(assetID)
+	if err != nil {
+		e.log.Panic("could not register taker fees, invalid asset", logging.Error(err))
+	}
+
+	assetQuantum := asset.Type().Details.Quantum
+
+	updatedKeys := map[partyAssetKey]struct{}{}
+	for party, fee := range feesPerParty {
+		key := e.feeDiscountKey(assetID, party)
+		updatedKeys[key] = struct{}{}
+
+		if _, ok := e.feeDiscountPerPartyAndAsset[key]; !ok {
+			e.feeDiscountPerPartyAndAsset[key] = num.UintZero()
+		}
+
+		// apply decay discount amount
+		e.feeDiscountPerPartyAndAsset[key] = e.decayFeeDiscountAmount(e.feeDiscountPerPartyAndAsset[key], assetQuantum)
+
+		// add fees
+		e.feeDiscountPerPartyAndAsset[key].Add(e.feeDiscountPerPartyAndAsset[key], fee)
+
+		updateDiscountEvents = append(updateDiscountEvents, events.NewTransferFeesDiscountUpdated(
+			ctx,
+			party,
+			assetID,
+			e.feeDiscountPerPartyAndAsset[key].Clone(),
+			e.currentEpoch,
+		))
+	}
+	for key := range e.feeDiscountPerPartyAndAsset {
+		if _, ok := updatedKeys[key]; ok {
+			continue
+		}
+
+		// apply decay discount amount
+		e.feeDiscountPerPartyAndAsset[key] = e.decayFeeDiscountAmount(e.feeDiscountPerPartyAndAsset[key], assetQuantum)
+
+		updateDiscountEvents = append(updateDiscountEvents, events.NewTransferFeesDiscountUpdated(
+			ctx,
+			key.party,
+			assetID,
+			e.feeDiscountPerPartyAndAsset[key].Clone(),
+			e.currentEpoch,
+		))
+	}
+
+	e.broker.SendBatch(updateDiscountEvents)
```

Comment:
I think `updateDiscountEvents` is populate from iterating over maps, so we should sort the events first so that the datanode history-segments don't diverge.